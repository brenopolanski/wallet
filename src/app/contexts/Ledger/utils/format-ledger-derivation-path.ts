import { LedgerDerivationScheme } from "app/contexts/Ledger/contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
