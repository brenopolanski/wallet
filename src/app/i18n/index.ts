import i18n from "i18next";
import { initReactI18next } from "react-i18next";

// Helpers
import { buildTranslations } from "./helpers";

const defaultNS = "translation";

const resources = {
	en: {
		translation: buildTranslations(),
	},
};

// eslint-disable-next-line
i18n.use(initReactI18next).init({
	defaultNS,
	lng: "en",
	ns: [defaultNS],
	resources,
});

export { defaultNS, i18n, resources };
