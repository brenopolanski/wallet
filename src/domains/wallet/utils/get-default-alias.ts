import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let sameCoinWalletsCount = sameCoinWallets.length || 1;

	while (profile.wallets().findByAlias(makeAlias(sameCoinWalletsCount))) {
		sameCoinWalletsCount++;
	}

	return makeAlias(sameCoinWalletsCount);
};
