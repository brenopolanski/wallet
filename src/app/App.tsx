// import { TRX } from "@payvo/sdk-trx";
// import { XLM } from "@payvo/sdk-xlm";
// import { XRP } from "@payvo/sdk-xrp";
// import { ZIL } from "@payvo/sdk-zil";
import "focus-visible";

import LedgerTransportNodeHID from "@ledgerhq/hw-transport-node-hid-singleton";
// import { LUNA } from "@payvo/sdk-luna";
// import { NANO } from "@payvo/sdk-nano";
// import { NEO } from "@payvo/sdk-neo";
import { Environment } from "@payvo/profiles";
// import { ADA } from "@payvo/sdk-ada";
import { ARK } from "@payvo/sdk-ark";
// import { ATOM } from "@payvo/sdk-atom";
// import { AVAX } from "@payvo/sdk-avax";
// import { BTC } from "@payvo/sdk-btc";
// import { DOT } from "@payvo/sdk-dot";
// import { EGLD } from "@payvo/sdk-egld";
// import { ETH } from "@payvo/sdk-eth";
import { LSK } from "@payvo/sdk-lsk";
import { Offline } from "domains/error/pages";
import { ExchangeProvider } from "domains/exchange/contexts/Exchange";
import { Splash } from "domains/splash/pages";
import React, { useLayoutEffect, useState } from "react";
import { useErrorHandler } from "react-error-boundary";
import { I18nextProvider, useTranslation } from "react-i18next";
import { useHistory } from "react-router-dom";
import { ToastContainer } from "react-toastify";
import { middlewares, RouterView, routes } from "router";
import { StubStorage } from "tests/mocks";
import { setThemeSource, shouldUseDarkColors } from "utils/electron-utils";
import { bootEnvWithProfileFixtures, isE2E, isUnit } from "utils/test-helpers";

import { SyncErrorMessage } from "./components/ProfileSyncStatusMessage";
import { ConfigurationProvider, EnvironmentProvider, LedgerProvider, useEnvironmentContext } from "./contexts";
import { useDeeplink, useNetworkStatus, useProfileSynchronizer } from "./hooks";
import { i18n as index18n } from "./i18n";
import { PluginProviders } from "./PluginProviders";
import { SentryProvider } from "./sentry/SentryProvider";
import { SentryRouterWrapper } from "./sentry/SentryRouterWrapper";
import { httpClient, toasts } from "./services";

const RouteWrappers = ({ children }: { children: React.ReactNode }) => (
	<SentryRouterWrapper>{children}</SentryRouterWrapper>
);

const Main = () => {
	const [showSplash, setShowSplash] = useState(true);
	const { env } = useEnvironmentContext();
	const isOnline = useNetworkStatus();
	const history = useHistory();
	const { t } = useTranslation();

	useProfileSynchronizer({
		onProfileRestoreError: () => history.push("/"),
		onProfileSignOut: () => {
			toasts.dismiss();
		},
		onProfileSyncComplete: async () => {
			await toasts.dismiss();
			toasts.success(t("COMMON.PROFILE_SYNC_COMPLETED"));
		},
		onProfileSyncError: async (failedNetworkNames, retryProfileSync) => {
			await toasts.dismiss();

			toasts.warning(
				<SyncErrorMessage
					failedNetworkNames={failedNetworkNames}
					onRetry={async () => {
						await toasts.dismiss();
						retryProfileSync();
					}}
				/>,
			);
		},
		onProfileSyncStart: () => {
			toasts.warning(t("COMMON.PROFILE_SYNC_STARTED"), { autoClose: false });
		},
	});

	useDeeplink();

	useLayoutEffect(() => {
		setThemeSource("system");

		document.body.classList.remove(`theme-${shouldUseDarkColors() ? "light" : "dark"}`);
		document.body.classList.add(`theme-${shouldUseDarkColors() ? "dark" : "light"}`);
	}, []);

	const handleError = useErrorHandler();

	useLayoutEffect(() => {
		const boot = async () => {
			try {
				/* istanbul ignore next */
				if (isE2E() || isUnit()) {
					await bootEnvWithProfileFixtures({ env, shouldRestoreDefaultProfile: isUnit() });

					setShowSplash(false);
					return;
				}

				/* istanbul ignore next */
				await env.verify();
				/* istanbul ignore next */
				await env.boot();
			} catch (error) {
				handleError(error);
			}

			setShowSplash(false);
		};

		boot();
	}, [env, handleError]);

	const renderContent = () => {
		if (showSplash) {
			return <Splash />;
		}

		if (!isOnline) {
			return <Offline />;
		}

		return <RouterView routes={routes} middlewares={middlewares} wrapper={RouteWrappers} />;
	};

	return (
		<main data-testid="Main">
			<ToastContainer closeOnClick={false} newestOnTop />

			{renderContent()}
		</main>
	);
};

export const App = () => {
	/**
	 * Ensure that the Environment object will not be recreated when the state changes,
	 * as the data is stored in memory by the `DataRepository`.
	 */

	/* istanbul ignore next */
	const storage = isE2E() || isUnit() ? new StubStorage() : "indexeddb";

	const [environment] = useState(
		() =>
			new Environment({
				coins: {
					// ADA,
					ARK,
					// ATOM,
					// AVAX,
					// BTC,
					// DOT,
					// ETH,
					// EGLD,
					LSK,
					// NEO,
					// NANO,
					// LUNA,
					// TRX,
					// XLM,
					// XRP,
					// ZIL,
				},
				httpClient,
				storage,
			}),
	);

	return (
		<I18nextProvider i18n={index18n}>
			<EnvironmentProvider env={environment}>
				<ConfigurationProvider defaultConfiguration={{ profileIsSyncingExchangeRates: true }}>
					<SentryProvider>
						<LedgerProvider transport={LedgerTransportNodeHID}>
							<PluginProviders>
								<ExchangeProvider>
									<Main />
								</ExchangeProvider>
							</PluginProviders>
						</LedgerProvider>
					</SentryProvider>
				</ConfigurationProvider>
			</EnvironmentProvider>
		</I18nextProvider>
	);
};
