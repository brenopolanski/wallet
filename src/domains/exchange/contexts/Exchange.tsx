import { camelCase } from "@payvo/sdk-helpers";
import { httpClient } from "app/services";
import { Exchange } from "domains/exchange/contracts";
import { exchangeHost, ExchangeService } from "domains/exchange/services/exchange.service";
import React, { useCallback, useMemo, useState } from "react";

interface Properties {
	children: React.ReactNode;
}

const ExchangeContext = React.createContext<any>(undefined);

const useExchange = () => {
	const [provider, setProvider] = useState<Exchange | undefined>();
	const [exchangeProviders, setExchangeProviders] = useState<Exchange[] | undefined>();

	const exchangeService = useMemo<ExchangeService | undefined>(() => {
		if (provider) {
			return new ExchangeService(provider.slug, httpClient);
		}
	}, [provider]);

	const fetchProviders = useCallback(async () => {
		const transformExchanges = (exchanges: Exchange[]) =>
			exchanges.map((exchange: Exchange) => {
				const newExchange: any = {};

				for (const [key, value] of Object.entries(exchange)) {
					newExchange[camelCase(key) as string] = value;
				}

				return newExchange;
			});

		try {
			const body = (await httpClient.get(exchangeHost)).json();
			setExchangeProviders(transformExchanges(body.data));
		} catch {
			setExchangeProviders([]);
		}
	}, []);

	return {
		exchangeProviders,
		exchangeService,
		fetchProviders,
		provider,
		setProvider,
	};
};

export const ExchangeProvider = ({ children }: Properties) => {
	const context = useExchange();

	return <ExchangeContext.Provider value={context}>{children}</ExchangeContext.Provider>;
};

export const useExchangeContext = (): ReturnType<typeof useExchange> => {
	const context = React.useContext(ExchangeContext);
	if (context === undefined) {
		throw new Error("[useExchangeContext] Component not wrapped within a Provider");
	}
	return context;
};
