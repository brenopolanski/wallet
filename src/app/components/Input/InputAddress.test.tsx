/* eslint-disable @typescript-eslint/require-await */
import { Contracts } from "@payvo/sdk-profiles";
import { renderHook } from "@testing-library/react-hooks";
import { EnvironmentProvider } from "app/contexts";
import { translations as commonTranslations } from "app/i18n/common/i18n";
import React from "react";
import { useForm } from "react-hook-form";
import { env, fireEvent, getDefaultProfileId, render, screen } from "utils/testing-library";

import { InputAddress, InputAddressProperties } from "./InputAddress";

let profile: Contracts.IProfile;

describe("InputAddress", () => {
	beforeAll(() => {
		profile = env.profiles().findById(getDefaultProfileId());
	});

	const TestInputAddress = (properties: InputAddressProperties) => (
		<EnvironmentProvider env={env}>
			<InputAddress name="address" {...properties} />
		</EnvironmentProvider>
	);

	it("should render", () => {
		const { asFragment } = render(
			<TestInputAddress coin="ARK" network="ark.devnet" profile={profile} />,
		);

		expect(screen.getByTestId("InputAddress__input")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should validate a wrong address", async () => {
		const { result, waitForNextUpdate } = renderHook(() => useForm({ mode: "onChange" }));
		const { register, errors } = result.current;

		render(
			<TestInputAddress coin="ARK" network="ark.devnet" registerRef={register} profile={profile} />,
		);

		fireEvent.input(screen.getByTestId("InputAddress__input"), { target: { value: "Abc" } });

		await waitForNextUpdate();

		expect(errors.address?.message).toBe(commonTranslations.INPUT_ADDRESS.VALIDATION.NOT_VALID);
	});

	it("should validate a valid address and emit event", async () => {
		const onValidAddress = jest.fn();
		const { result, waitForNextUpdate } = renderHook(() => useForm({ mode: "onChange" }));
		const { register, errors } = result.current;
		const validAddress = "DT11QcbKqTXJ59jrUTpcMyggTcwmyFYRTM";

		render(
			<TestInputAddress
				coin="ARK"
				network="ark.devnet"
				registerRef={register}
				onValidAddress={onValidAddress}
				profile={profile}
			/>,
		);

		fireEvent.input(screen.getByTestId("InputAddress__input"), {
			target: { value: validAddress },
		});

		await waitForNextUpdate();

		expect(errors.address?.message).toBeUndefined();
		expect(onValidAddress).toHaveBeenCalledWith(validAddress);
	});

	it("should validate with additional rules", async () => {
		const { result, waitForNextUpdate } = renderHook(() => useForm({ mode: "onChange" }));
		const { register, errors } = result.current;

		render(
			<TestInputAddress
				profile={profile}
				coin="ARK"
				network="ark.devnet"
				registerRef={register}
				additionalRules={{ minLength: 10 }}
			/>,
		);

		fireEvent.input(screen.getByTestId("InputAddress__input"), { target: { value: "Abc" } });

		await waitForNextUpdate();

		expect(errors.address?.type).toBe("minLength");
	});

	it("should not use default validation", async () => {
		const { result } = renderHook(() => useForm({ mode: "onChange" }));
		const { register } = result.current;

		const { asFragment } = render(
			<TestInputAddress useDefaultRules={false} registerRef={register} profile={profile} />,
		);

		expect(asFragment()).toMatchSnapshot();
	});
});
