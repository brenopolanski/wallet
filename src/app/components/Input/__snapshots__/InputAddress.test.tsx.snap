// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputAddress should not use default validation 1`] = `
<DocumentFragment>
  <div
    class="Input__InputWrapperStyled-sc-1i8sd1r-0 cLNVAR"
  >
    <div
      class="relative flex flex-1 h-full"
    >
      <input
        class="Input__InputStyled-sc-1i8sd1r-1 hPntwa p-0 border-none bg-transparent focus:ring-0 no-ligatures w-full"
        data-testid="InputAddress__input"
        name="address"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputAddress should render 1`] = `
<DocumentFragment>
  <div
    class="Input__InputWrapperStyled-sc-1i8sd1r-0 cLNVAR"
  >
    <div
      class="relative flex flex-1 h-full"
    >
      <input
        class="Input__InputStyled-sc-1i8sd1r-1 hPntwa p-0 border-none bg-transparent focus:ring-0 no-ligatures w-full"
        data-testid="InputAddress__input"
        name="address"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
